File: Navigation/CrumbTrailManager.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Text;
using System.Text.RegularExpressions;
using System.Web;

namespace Navigation
{
	internal static class CrumbTrailManager
	{
		private static string SEPARATOR = !NavigationSettings.Config.OriginalUrlSeparators ? "_" : "!";
		private static string RET_1_SEP = "1" + SEPARATOR;
		private static string RET_2_SEP = "2" + SEPARATOR;
		private static string RET_3_SEP = "3" + SEPARATOR;
		private static string CRUMB_1_SEP = "4" + SEPARATOR;
		private static string CRUMB_2_SEP = "5" + SEPARATOR;

		internal static List<Crumb> Crumbs
		{
			get
			{
				List<Crumb> crumbTrailArray = new List<Crumb>();
				int arrayCount = 0;
				string trail = StateContext.CrumbTrail;
				int crumbTrailSize = trail == null ? 0 : Regex.Split(trail, CRUMB_1_SEP).Length - 1;
				NavigationData navigationData;
				bool last = true;
				string stateKey = null;
				State state = null;
				while (arrayCount < crumbTrailSize)
				{
					stateKey = Regex.Split(trail.Substring(CRUMB_1_SEP.Length), CRUMB_2_SEP)[0];
					state = StateContext.GetState(stateKey);
					navigationData = GetCrumbTrailData(trail, state);
					trail = CropCrumbTrail(trail);
					crumbTrailArray.Add(new Crumb(navigationData, state, last));
					last = false;
					arrayCount++;
				}
				crumbTrailArray.Reverse();
				return crumbTrailArray;
			}
		}

		private static NavigationData GetCrumbTrailData(string trail, State state)
		{
			NavigationData navData = null;
			string data = Regex.Split(trail.Substring(trail.IndexOf(CRUMB_2_SEP, StringComparison.Ordinal) + CRUMB_2_SEP.Length), CRUMB_1_SEP)[0];
			if (data.Length != 0)
				navData = ParseReturnData(data, state);
			return navData;
		}

		private static string CropCrumbTrail(string trail)
		{
			string croppedTrail;
			int nextTrailStart = trail.IndexOf(CRUMB_1_SEP, 1, StringComparison.Ordinal);
			if (nextTrailStart != -1)
				croppedTrail = trail.Substring(nextTrailStart);
			else
				croppedTrail = "";
			return croppedTrail;
		}

		internal static void BuildCrumbTrail()
		{
			var crumbs = Crumbs;
			if (StateContext.PreviousState != null)
				crumbs.Add(new Crumb(StateContext.ReturnData, StateContext.PreviousState, false));
			crumbs = StateContext.State.StateHandler.TruncateCrumbTrail(StateContext.State, crumbs);
			crumbs.Reverse();
			var trailBuilder = new StringBuilder();
			foreach (var crumb in crumbs)
			{
				trailBuilder.Append(CRUMB_1_SEP);
				trailBuilder.Append(crumb.State.StateKey);
				trailBuilder.Append(CRUMB_2_SEP);
				FormatReturnData(trailBuilder, crumb.State, crumb.Data);
			}
			StateContext.GenerateKey(trailBuilder.Length != 0 ? trailBuilder.ToString() : null);
		}

		internal static string GetHref(string nextState, NavigationData navigationData, NavigationData returnData)
		{
			string previousState = StateContext.StateId;
			string crumbTrail = StateContext.CrumbTrailKey;
			State state = StateContext.GetState(nextState);
			NameValueCollection coll = new NameValueCollection();
			coll[NavigationSettings.Config.StateIdKey] = nextState;
			if (previousState != null && state.TrackCrumbTrail)
			{
				coll[NavigationSettings.Config.PreviousStateIdKey] = previousState;
			}
			if (navigationData != null)
			{
				foreach (NavigationDataItem item in navigationData)
				{
					if (!item.Value.Equals(string.Empty) && !state.DefaultOrDerived(item.Key, item.Value))
						coll[item.Key] = FormatURLObject(item.Key, item.Value, state);
				}
			}
			if (returnData != null && state.TrackCrumbTrail && StateContext.State != null)
			{
				var returnDataBuilder = FormatReturnData(null, StateContext.State, returnData);
				if (returnDataBuilder.Length > 0)
					coll[NavigationSettings.Config.ReturnDataKey] = returnDataBuilder.ToString();
			}
			if (crumbTrail != null && state.TrackCrumbTrail)
			{
				coll[NavigationSettings.Config.CrumbTrailKey] = crumbTrail;
			}
#if NET35Plus
			coll = StateContext.ShieldEncode(coll, false, state);
#else
			coll = StateContext.ShieldEncode(coll, state);
#endif
#if NET40Plus
			HttpContextBase context = null;
			if (HttpContext.Current != null)
				context = new HttpContextWrapper(HttpContext.Current);
			else
				context = new MockNavigationContext(null, state);
			return state.StateHandler.GetNavigationLink(state, coll, context);
#else
			return state.StateHandler.GetNavigationLink(state, coll);
#endif
		}

		private static StringBuilder FormatReturnData(StringBuilder returnDataBuilder, State state, NavigationData returnData)
		{
			returnDataBuilder = returnDataBuilder ?? new StringBuilder();
			string prefix = string.Empty;
			foreach (NavigationDataItem item in returnData)
			{
				if (!item.Value.Equals(string.Empty) && !state.DefaultOrDerived(item.Key, item.Value))
				{
					returnDataBuilder.Append(prefix);
					returnDataBuilder.Append(EncodeURLValue(item.Key));
					returnDataBuilder.Append(RET_1_SEP);
					returnDataBuilder.Append(FormatURLObject(item.Key, item.Value, state));
					prefix = RET_3_SEP;
				}
			}
			return returnDataBuilder;
		}

		private static string DecodeURLValue(string urlValue)
		{
			return urlValue.Replace("0" + SEPARATOR, SEPARATOR);
		}

		private static string EncodeURLValue(string urlValue)
		{
			return urlValue.Replace(SEPARATOR, "0" + SEPARATOR);
		}

		internal static string FormatURLObject(string key, object urlObject, State state)
		{
			Type defaultType = state.DefaultTypes[key] ?? typeof(string);
			string converterKey = ConverterFactory.GetKey(urlObject);
			string formattedValue = ConverterFactory.GetConverter(converterKey).ConvertToInvariantString(urlObject);
			formattedValue = EncodeURLValue(formattedValue);
			if (urlObject.GetType() != defaultType)
				formattedValue += RET_2_SEP + converterKey;
			return formattedValue;
		}

		internal static object ParseURLString(string key, string val, State state)
		{
			Type defaultType = state.DefaultTypes[key] ?? typeof(string);
			string urlValue = val;
			string converterKey = ConverterFactory.GetKey(defaultType);
			if (val.IndexOf(RET_2_SEP, StringComparison.Ordinal) > -1)
			{
				string[] arr = Regex.Split(val, RET_2_SEP);
				urlValue = arr[0];
				converterKey = arr[1];
			}
			try
			{
				return ConverterFactory.GetConverter(converterKey).ConvertFromInvariantString(DecodeURLValue(urlValue));
			}
			catch (Exception ex)
			{
				throw new UrlException(Resources.InvalidUrl, ex);
			}
		}

		internal static string GetRefreshHref(NavigationData refreshData)
		{
			return GetHref(StateContext.StateId, refreshData, null);
		}

		internal static object Parse(string key, string val, State state)
		{
			object parsedVal;
			if (key == NavigationSettings.Config.ReturnDataKey)
			{
				parsedVal = ParseReturnData(val, state);
			}
			else
			{
				if (key == NavigationSettings.Config.CrumbTrailKey)
				{
					parsedVal = val;
				}
				else
				{
					parsedVal = ParseURLString(key, val, state);
				}
			}
			return parsedVal;
		}

		private static NavigationData ParseReturnData(string returnData, State state)
		{
			NavigationData navData = new NavigationData();
			string[] nameValuePair;
			string[] returnDataArray = Regex.Split(returnData, RET_3_SEP);
			for (int i = 0; i < returnDataArray.Length; i++)
			{
				nameValuePair = Regex.Split(returnDataArray[i], RET_1_SEP);
				navData.Add(DecodeURLValue(nameValuePair[0]), ParseURLString(DecodeURLValue(nameValuePair[0]), nameValuePair[1], state));
			}
			return navData;
		}
	}
}

File: NavigationJS/src/CrumbTrailManager.ts
import ConverterFactory = require('./converter/ConverterFactory');
import Crumb = require('./config/Crumb');
import Dialog = require('./config/Dialog');
import NavigationData = require('./NavigationData');
import NavigationSettings = require('./NavigationSettings');
import ReturnDataManager = require('./ReturnDataManager');
import State = require('./config/State');
import StateContext = require('./StateContext');

class CrumbTrailManager {
    //private static CRUMB_1_SEP = '4_';
    //private static CRUMB_2_SEP = '5_';

    static buildCrumbTrail(stateContext: StateContext, settings: NavigationSettings, converterFactory: ConverterFactory, dialogs: Dialog[], uncombined: boolean) {
        var crumbs = this.getCrumbs(stateContext, settings, converterFactory, dialogs, false);
        crumbs = stateContext.state.stateHandler.truncateCrumbTrail(stateContext.state, crumbs);
        stateContext.crumbTrail = [];
        for(var i = 0; i < crumbs.length; i++)
            stateContext.crumbTrail.push(this.removeCrumbs(crumbs[i].navigationLink));
        stateContext.crumbTrail.push(this.removeCrumbs(stateContext.url));
        /*var crumbs = this.getCrumbs(stateContext, settings, converterFactory, dialogs, false);
        if (uncombined)
            crumbs.push(this.getCrumb(stateContext, settings, converterFactory, stateContext.previousState, stateContext.previousData, false));        
        crumbs = stateContext.state.stateHandler.truncateCrumbTrail(stateContext.state, crumbs);
        if (settings.combineCrumbTrail)
            crumbs.push(this.getCrumb(stateContext, settings, converterFactory, stateContext.state, stateContext.data, false));
        crumbs.reverse();
        var trailString: string = '';
        for (var i = 0; i < crumbs.length; i++) {
            trailString += this.CRUMB_1_SEP + crumbs[i].state.id + this.CRUMB_2_SEP;
            trailString += ReturnDataManager.formatReturnData(settings, converterFactory, crumbs[i].state, crumbs[i].data);
        }
        stateContext.crumbTrail = trailString ? trailString : null;
        stateContext.crumbTrailKey = settings.crumbTrailPersister.save(stateContext.crumbTrail, stateContext.state);*/
    }

    static getCrumbs(stateContext: StateContext, settings: NavigationSettings, converterFactory: ConverterFactory, dialogs: Dialog[], setLast: boolean, skipLatest?: boolean): Crumb[] {
        var crumbTrailArray: Crumb[] = [];
        var len = stateContext.crumbTrail.length - (skipLatest ? 1 : 0);
        for(var i = 0; i < len; i++) {
            var state = settings.router.getData(stateContext.crumbTrail[i].split('?')[0]).state;
            crumbTrailArray.push(new Crumb(null, state, stateContext.crumbTrail[i], i === len - 1));            
        }
        //console.log(crumbTrailArray);
        /*var arrayCount = 0;
        var trail = stateContext.crumbTrail;
        var crumbTrailSize = !trail ? 0 : trail.split(this.CRUMB_1_SEP).length - 1;
        var last = true;
        while (arrayCount < crumbTrailSize) {
            var stateKey = trail.substring(this.CRUMB_1_SEP.length).split(this.CRUMB_2_SEP)[0];
            var state = this.getState(stateKey, dialogs);
            var navigationData: any = {};
            var data = trail.substring((trail.indexOf(this.CRUMB_2_SEP) + this.CRUMB_2_SEP.length)).split(this.CRUMB_1_SEP)[0];
            if (data)
                navigationData = ReturnDataManager.parseReturnData(settings, converterFactory, data, state);
            var nextTrailStart = trail.indexOf(this.CRUMB_1_SEP, 1);
            trail = nextTrailStart != -1 ? trail.substring(nextTrailStart) : '';
            if (!skipLatest) {
                crumbTrailArray.push(this.getCrumb(stateContext, settings, converterFactory, state, navigationData, setLast && last));
                last = false;
            }
            skipLatest = false;
            arrayCount++;
        }
        crumbTrailArray.reverse();*/
        return crumbTrailArray;
    }
    
    static getCrumb(stateContext: StateContext, settings: NavigationSettings, converterFactory: ConverterFactory, state: State, data: any, last: boolean): Crumb {
        var link = this.getHref(stateContext, settings, converterFactory, state, data, null);
        return new Crumb(NavigationData.setDefaults(data, state.defaults), state, link, last);        
    }

    /*static getState(id: string, dialogs: Dialog[]) {
        if (!id) return null;
        var ids = id.split('-');
        return dialogs[+ids[0]]._states[+ids[1]];
    }*/

    static getHref(stateContext: StateContext, settings: NavigationSettings, converterFactory: ConverterFactory, state: State, navigationData: any, returnData: any): string {
        var data = {};
        /*if (!settings.combineCrumbTrail && state.trackCrumbTrail && stateContext.state)
            data[settings.previousStateIdKey] = stateContext.state.id;*/
        if (!settings.router.supportsDefaults) {
            navigationData = NavigationData.clone(navigationData);
            NavigationData.setDefaults(navigationData, state.defaults);
        }
        var arrayData: { [index: string]: string[] } = {};
        for (var key in navigationData) {
            var val = navigationData[key]; 
            if (val != null && val.toString()) {
                var formattedData = ReturnDataManager.formatURLObject(settings, converterFactory, key, val, state);
                val = formattedData.val;
                if (!settings.router.supportsDefaults || val !== state.formattedDefaults[key]) {
                    data[key] = val;
                    arrayData[key] = formattedData.arrayVal;
                }
            }
        }
        /*if (state.trackCrumbTrail && stateContext.crumbTrail.length > 0) {
            var formattedCrumbData = ReturnDataManager.formatURLObject(settings, converterFactory, 'crumb', stateContext.crumbTrail, state);
            var crumbVal = formattedCrumbData.val;
            data['crumb'] = crumbVal;
            arrayData['crumb'] = formattedCrumbData.arrayVal;
        }*/
        /*if (!settings.combineCrumbTrail && state.trackCrumbTrail && stateContext.state) {
            if (settings.trackAllPreviousData)
                returnData = stateContext.data;
            var returnDataString = ReturnDataManager.formatReturnData(settings, converterFactory, stateContext.state, returnData);
            if (returnDataString)
                data[settings.returnDataKey] = returnDataString;
        }*
        if (stateContext.crumbTrailKey && state.trackCrumbTrail)
            data[settings.crumbTrailKey] = stateContext.crumbTrailKey;*/
        var link = state.stateHandler.getNavigationLink(settings.router, state, data, arrayData);
        if (state.trackCrumbTrail)
            link = this.appendCrumbs(link, stateContext.crumbTrail);
        return link;
    }
    
    static removeCrumbs(link: string): string{
        var ind = link.indexOf('crumb=');
        if (ind >= 0)
            link = link.substring(0, ind - 1);
        return link;
    }
    
    static appendCrumbs(link: string, crumbs: string[]): string {
        var sep = link.indexOf('?') >= 0 ? '&' : '?';
        for(var i = 0; i < crumbs.length; i++) {
            link += sep + 'crumb=' + encodeURIComponent(crumbs[i]);
            sep = '&';
        }
        return link;
    }

    static getRefreshHref(stateContext: StateContext, settings: NavigationSettings, converterFactory: ConverterFactory, refreshData: any): string {
        return this.getHref(stateContext, settings, converterFactory, stateContext.state, refreshData, null);
    }
}
export = CrumbTrailManager;
